# Genomic molecular characterization for viral strains using informatics tools

## CGS, USAMRIID Viral Population Genomics Training Module

## Objective:  
The training module will provide the complete bioinformatics workflow for analyzing genomics data using open source tools. The training module uses sequence reads generated using genomics tools such as genomic DNA or RNA sequencing using next generation sequencing technology with objective of characterization of viral strains in outbreak setting. 
## Introduction:
Infectious diseases are the most common cause of major fatalities of human life. Ebola virus disease (EVD) is one of the deadliest disease and observed in several outbreaks occurred primarily on the African continent. EVD effects both non-human primates (NHPs) and humans. It is the infection caused by the group of Ebola virus of family Filoviridae. There mainly four viruses Ebola, Sudan, Ta√Ø Forest, and Bundibugyo viruses which causes deadly disease in people. Viruses tend to evolve very fast with DNA mutations in their smaller genomes. In epidemiological studies often in outbreak setting genomics tools are often used in studying the specific mutations in viral strains and sub-population that can potentially make these viruses more resistant to treatments. Early during the recent Ebola epidemic in West Africa, scientists speculated that the genetic diversity of the circulating Makona strain of virus (EBOV-Makona) would result in more severe disease and more transmissibility than prior strains, as there are mutations observed on the viral surface glycoprotein. The training module will use the data generated in-house for viral plasmid EBOV Makona virus strain (SL3864-8U) closest to GenBank match Zaire ebolavirus isolate Ebola virus/H.sapiens-wt/SLE/2014/Makona-G3677.2, complete genome (KM034557.1). We will be using this as a reference genome to study the reported mutations in specific plasmids (414 (SL3864), 424 (NP UTR), 440 (GP V82A). In this module we will use open source bioinformatics tool to specific mutation in three other plasmids from the recovered viral isolates. 

## For cgs-popgen-2020 Docker Image build follow instructions

## Getting Started

These instructions will cover usage information and for the docker container 

## For downloading docker container please use following URL and instructions
*[Docker](https://hub.docker.com/repository/docker/rkumar23/cgs-popgen-2020)

## Find Us

* [GitHub](https://github.com/your/repository)
* [Quay.io](https://quay.io/repository/your/docker-repository)

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the 
[tags on this repository](https://github.com/your/repository/tags). 

## Authors

* **Raina Kumar** - *Initial work* - [rainakumar](https://github.com/rainakumar/CGS_popgen_training_module)

See also the list of [contributors](https://github.com/your/repository/contributors) who 
participated in this project.

* Joshua Richardson
* Nick Rozak
* MAJ Jeffery Kugelman

## License

This project is licensed under the Open Source License - see the [LICENSE.md](LICENSE.md) file for details.

## Acknowledgments

* People you want to thank
* If you took a bunch of code from somewhere list it here
